### ISSUE TITLE
Partial types not detected as protocol

### ISSUE THREAD

**Issue by marvinvanaalst** [microsoft/pyright#10289]
Type: <b>Bug</b>

When one uses the `type` of a `Protocol` as an input (e.g. to defer instantiation of a class because some other information might need to be passed), pylance trips on partial function application of the constructor, e.g. `partial(A, default_value=1)`. While this clearly also returns the constructor `A`, which adheres to the protocol, the inferred type `partial[A]` is not recognised as also adhering to the protocol.

```python
from dataclasses import dataclass
from functools import partial
from typing import Protocol


class MyProtocol(Protocol):
    def some_fn(self) -> int: ...


@dataclass
class A:
    default_value: int = 0

    def some_fn(self) -> int:
        return self.default_value + 1


@dataclass
class B:
    default_value: int = 0

    def some_fn(self) -> int:
        return self.default_value + 1


def fn(x: type[MyProtocol]) -> int:
    # Let's imagine for some reason we need to initialise x here
    return x().some_fn()


fn(A)  # no issue here
fn(B)  # no issue here

# Let's say we need to change the default value
fn(partial(A, default_value=1))  # why are these errors?
fn(partial(B, default_value=1))  # why are these errors?
```

Extension version: 2025.4.1
VS Code version: Code 1.99.0 (4437686ffebaf200fa4a6e6e67f735f3edf24ada, 2025-04-02T21:35:19.530Z)
OS version: Linux x64 6.13.9-200.fc41.x86_64
Modes:

<details>
<summary>System Added Logs</summary>

```
2025-04-10 08:39:50.459 [info] (Client) Pylance async client (2025.4.1) started with python extension (2025.4.0)
2025-04-10 08:39:51.175 [info] [Info  - 8:39:51 AM] (83785) Server root directory: file:///home/marvin/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist
2025-04-10 08:39:51.176 [info] [Info  - 8:39:51 AM] (83785) Pylance language server 2025.4.1 (pyright version 1.1.398, commit 4f24eccc) starting
2025-04-10 08:39:51.182 [info] [Info  - 8:39:51 AM] (83785) Starting service instance "ChillSugar" for workspace "/home/marvin/git/ChillSugar"
2025-04-10 08:39:51.251 [info] [Info  - 8:39:51 AM] (83785) Setting environmentName for service "ChillSugar": "3.13.2 (global)"
2025-04-10 08:39:51.252 [info] [Info  - 8:39:51 AM] (83785) Setting pythonPath for service "ChillSugar": "/bin/python"
2025-04-10 08:39:51.252 [info] [Info  - 8:39:51 AM] (83785) No include entries specified; assuming /home/marvin/git/ChillSugar
2025-04-10 08:39:51.253 [info] [Info  - 8:39:51 AM] (83785) Auto-excluding **/node_modules
2025-04-10 08:39:51.253 [info] [Info  - 8:39:51 AM] (83785) Auto-excluding **/__pycache__
2025-04-10 08:39:51.253 [info] [Info  - 8:39:51 AM] (83785) Auto-excluding **/.*
2025-04-10 08:39:51.290 [info] [Info  - 8:39:51 AM] (83785) Assuming Python version 3.13.2.final.0
2025-04-10 08:39:51.380 [info] [Info  - 8:39:51 AM] (83785) Found 1 source file
2025-04-10 08:39:51.404 [info] [Info  - 8:39:51 AM] (83785) Setting environmentName for service "ChillSugar": "3.13.2 (global)"
2025-04-10 08:39:51.404 [info] [Info  - 8:39:51 AM] (83785) Setting pythonPath for service "ChillSugar": "/bin/python"
2025-04-10 08:39:51.404 [info] [Info  - 8:39:51 AM] (83785) No include entries specified; assuming /home/marvin/git/ChillSugar
2025-04-10 08:39:51.404 [info] [Info  - 8:39:51 AM] (83785) Auto-excluding **/node_modules
2025-04-10 08:39:51.404 [info] [Info  - 8:39:51 AM] (83785) Auto-excluding **/__pycache__
2025-04-10 08:39:51.404 [info] [Info  - 8:39:51 AM] (83785) Auto-excluding **/.*
2025-04-10 08:39:51.444 [info] [Info  - 8:39:51 AM] (83785) Assuming Python version 3.13.2.final.0
2025-04-10 08:39:51.503 [info] [Info  - 8:39:51 AM] (83785) Found 1 source file
2025-04-10 08:39:51.922 [info] [Info  - 8:39:51 AM] (83785) BG: Priority queue background worker(2) root directory: file:///home/marvin/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist
2025-04-10 08:39:51.922 [info] [Info  - 8:39:51 AM] (83785) BG: Priority queue background worker(2) started
2025-04-10 08:43:52.805 [info] [Info  - 8:43:52 AM] (83785) BG: Indexer background runner(3) root directory: file:///home/marvin/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist (index)
2025-04-10 08:43:52.805 [info] [Info  - 8:43:52 AM] (83785) BG: Indexing(3) started
2025-04-10 08:43:53.080 [info] [Info  - 8:43:53 AM] (83785) BG: scanned(3) 96 files over 1 exec env
2025-04-10 08:43:53.740 [info] [Info  - 8:43:53 AM] (83785) BG: indexed(3) 96 files over 1 exec env
2025-04-10 08:43:53.772 [info] [Info  - 8:43:53 AM] (83785) BG: Indexing finished(3).
2025-04-10 08:43:55.654 [info] [Info  - 8:43:55 AM] (83785) Starting service instance "ChillSugar" for workspace "/home/marvin/git/ChillSugar"
2025-04-10 08:43:55.748 [info] [Info  - 8:43:55 AM] (83785) Setting environmentName for service "ChillSugar": "3.13.2 (global)"
2025-04-10 08:43:55.748 [info] [Info  - 8:43:55 AM] (83785) Setting pythonPath for service "ChillSugar": "/home/marvin/git/ChillSugar"
2025-04-10 08:43:55.748 [info] [Info  - 8:43:55 AM] (83785) No include entries specified; assuming /home/marvin/git/ChillSugar
2025-04-10 08:43:55.748 [info] [Info  - 8:43:55 AM] (83785) Auto-excluding **/node_modules
2025-04-10 08:43:55.748 [info] [Info  - 8:43:55 AM] (83785) Auto-excluding **/__pycache__
2025-04-10 08:43:55.748 [info] [Info  - 8:43:55 AM] (83785) Auto-excluding **/.*
2025-04-10 08:43:55.786 [info] [Info  - 8:43:55 AM] (83785) Found 1 source file
2025-04-10 08:43:56.472 [info] [Info  - 8:43:56 AM] (83785) BG: Priority queue background worker(5) root directory: file:///home/marvin/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist
2025-04-10 08:43:56.473 [info] [Info  - 8:43:56 AM] (83785) BG: Priority queue background worker(5) started
2025-04-10 08:43:56.603 [info] [Info  - 8:43:56 AM] (83785) BG: Indexer background runner(6) root directory: file:///home/marvin/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist (index)
2025-04-10 08:43:56.603 [info] [Info  - 8:43:56 AM] (83785) BG: Indexing(6) started
2025-04-10 08:43:56.854 [info] [Info  - 8:43:56 AM] (83785) BG: scanned(6) 1 files over 1 exec env
2025-04-10 08:43:56.991 [info] [Info  - 8:43:56 AM] (83785) BG: indexed(6) 1 files over 1 exec env
2025-04-10 08:43:56.998 [info] [Info  - 8:43:56 AM] (83785) BG: Indexing finished(6).
2025-04-10 08:44:22.354 [info] [Info  - 8:44:22 AM] (83785) Starting service instance "ChillSugar" for workspace "/home/marvin/git/ChillSugar"
2025-04-10 08:44:22.430 [info] [Info  - 8:44:22 AM] (83785) Setting environmentName for service "ChillSugar": "3.13.2 (global)"
2025-04-10 08:44:22.432 [info] [Info  - 8:44:22 AM] (83785) Setting pythonPath for service "ChillSugar": "/home/marvin/git/ChillSugar/.pixi/envs/default/bin/python"
2025-04-10 08:44:22.432 [info] [Info  - 8:44:22 AM] (83785) No include entries specified; assuming /home/marvin/git/ChillSugar
2025-04-10 08:44:22.432 [info] [Info  - 8:44:22 AM] (83785) Auto-excluding **/node_modules
2025-04-10 08:44:22.433 [info] [Info  - 8:44:22 AM] (83785) Auto-excluding **/__pycache__
2025-04-10 08:44:22.433 [info] [Info  - 8:44:22 AM] (83785) Auto-excluding **/.*
2025-04-10 08:44:22.441 [info] [Info  - 8:44:22 AM] (83785) Assuming Python version 3.13.2.final.0
2025-04-10 08:44:22.576 [info] [Info  - 8:44:22 AM] (83785) Found 1 source file
2025-04-10 08:44:23.187 [info] [Info  - 8:44:23 AM] (83785) BG: Priority queue background worker(8) root directory: file:///home/marvin/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist
2025-04-10 08:44:23.188 [info] [Info  - 8:44:23 AM] (83785) BG: Priority queue background worker(8) started
2025-04-10 08:44:23.261 [info] [Info  - 8:44:23 AM] (83785) BG: Indexer background runner(9) root directory: file:///home/marvin/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist (refresh)
2025-04-10 08:44:23.262 [info] [Info  - 8:44:23 AM] (83785) BG: Indexing(9) started
2025-04-10 08:44:23.923 [info] [Info  - 8:44:23 AM] (83785) BG: scanned(9) 197 files over 1 exec env
2025-04-10 08:44:26.569 [info] [Info  - 8:44:26 AM] (83785) BG: [IDX(9)] Long operation: index execution environment file:///home/marvin/git/ChillSugar (2465ms)
2025-04-10 08:44:26.591 [info] [Info  - 8:44:26 AM] (83785) BG: [IDX(9)] Long operation: index packages file:///home/marvin/git/ChillSugar (2515ms)
2025-04-10 08:44:26.591 [info] [Info  - 8:44:26 AM] (83785) BG: indexed(9) 197 files over 1 exec env
2025-04-10 08:44:26.661 [info] [Info  - 8:44:26 AM] (83785) BG: Indexing finished(9).
2025-04-10 08:45:59.357 [info] [Info  - 8:45:59 AM] (83785) Starting service instance "<default>" for workspace ""
2025-04-10 08:45:59.418 [info] [Info  - 8:45:59 AM] (83785) Setting environmentName for service "<default>": "3.13.2 (global)"
2025-04-10 08:45:59.421 [info] [Info  - 8:45:59 AM] (83785) Setting pythonPath for service "<default>": "/bin/python"
2025-04-10 08:45:59.422 [info] [Info  - 8:45:59 AM] (83785) No include entries specified; assuming /<default workspace root>
2025-04-10 08:45:59.422 [info] [Info  - 8:45:59 AM] (83785) Auto-excluding **/node_modules
2025-04-10 08:45:59.422 [info] [Info  - 8:45:59 AM] (83785) Auto-excluding **/__pycache__
2025-04-10 08:45:59.422 [info] [Info  - 8:45:59 AM] (83785) Auto-excluding **/.*
2025-04-10 08:45:59.440 [info] [Info  - 8:45:59 AM] (83785) Assuming Python version 3.13.2.final.0
2025-04-10 08:45:59.518 [info] [Error - 8:45:59 AM] (83785) File or directory "/<default workspace root>" does not exist.
2025-04-10 08:45:59.518 [info] [Info  - 8:45:59 AM] (83785) No source files found.
2025-04-10 08:46:00.145 [info] [Info  - 8:46:00 AM] (83785) BG: Priority queue background worker(11) root directory: file:///home/marvin/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist
2025-04-10 08:46:00.146 [info] [Info  - 8:46:00 AM] (83785) BG: Priority queue background worker(11) started
2025-04-10 08:58:03.377 [info] [Info  - 8:58:03 AM] (83785) Found 13 source files
2025-04-10 08:58:03.406 [info] [Info  - 8:58:03 AM] (83785) Found 11 source files
2025-04-10 08:58:03.598 [info] [Info  - 8:58:03 AM] (83785) Found 2 source files

```
</details><details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) Ultra 7 165H (22 x 1791)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off<br>webnn: disabled_off|
|Load (avg)|0, 1, 1|
|Memory (System)|30.79GB (19.78GB free)|
|Process Argv|--crash-reporter-id 37f579ff-5932-40a7-a6af-879517e89042|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|gnome|
|XDG_CURRENT_DESKTOP|GNOME|
|XDG_SESSION_DESKTOP|gnome|
|XDG_SESSION_TYPE|wayland|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805:30301674
binariesv615:30325510
vscaac:30438847
c4g48928:30535728
azure-dev_surveyone:30548225
962ge761:30959799
2e7ec940:31000449
pythontbext0:30879054
cppperfnew:31000557
dwnewjupytercf:31046870
pythonrstrctxt:31112756
nativeloc1:31192215
5fd0e150:31155592
dwcopilot:31170013
6074i472:31201624
dwoutputs:31242946
customenabled:31248079
9064b325:31222308
copilot_t_ci:31222730
fiia8375:31280369
pythoneinst12cf:31262606
bgtreat:31268568
4gafe986:31271826
31787653:31262186
3e8i5726:31271747
996jf627:31281244
useunpkgapi:31280918
7bj51361:31279165
747dc170:31275177
aj496949:31278748
aj953862:31281341

```

</details>

**Comment by marvinvanaalst** [comment]
Ah, thanks for clearing this up. It *feels* like `partial` shouldn't work that way, but it is what it is 😄 
How can we fix the code then?
My first idea would be to use `Callable` instead, since that's what a constructor is in the end. 

```python
def fn(x: Callable[[], MyProtocol]) -> int:
    return x().some_fn()
```

That does seem to work so far.
Any other / better suggestion of how to write this in a type-safe way?

**Comment by erictraut** [comment]
Yes, a `Callable` is the mechanism I'd suggest here.
